File: root/scala/batch_app.scala
import org.apache.spark.sql.{SparkSession, Row}
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._


val spark = SparkSession.builder.appName("batch-app").getOrCreate()

// Define HDFS paths for RAW and Processed Zones
val hdfsRawPath = "hdfs://user/data/raw_zone"
val hdfsProcessedPath = "hdfs://user/data/processed_zone"

// Define schema for the data in RAW Zone
val rawSchema = StructType(Seq(
  StructField("name", StringType),
  StructField("date", StringType),
  StructField("sever-details", StructType(Seq(
    StructField("server_id", StringType),
    StructField("location", StringType),
    StructField("temp", StringType)
  )))
))

// Read data from RAW Zone
val rawData = spark.read.schema(rawSchema).parquet(hdfsRawPath)

// Calculate the timestamp for the current hour - 1
val currentDateTimeMinus1 = current_timestamp() - hours(1)

// Filter data for the last day
val incrementalData = rawData.filter(unix_timestamp(col("date")) > currentDateTimeMinus1)

val processedData = incrementalData.select(
  col("name"),
  col("date"),
  col("sever-details.server_id").cast(IntegerType).alias("server_id"),
  col("sever-details.location"),
  col("sever-details.temp")
)

// Append processed data in the Processed Zone
processedData.write.mode("append").partitionBy("date").parquet(hdfsProcessedPath)

spark.stop()

File: root/src/batch_app.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, unix_timestamp, current_timestamp
from pyspark.sql.types import StringType, StructType, StructField

# Create a Spark Session
spark = SparkSession.builder.appName("IncrementalLoadJob").getOrCreate()

# Define HDFS paths for RAW and Processed Zones
hdfs_raw_path = "hdfs://user/data/raw_zone"
hdfs_processed_path = "hdfs://user/data/processed_zone"

# Define schema for the data in RAW Zone
raw_schema = StructType([
    StructField("name", StringType()),
    StructField("date", StringType()),
    StructField("sever-details", StructType([
        StructField("server_id", StringType()),
        StructField("location", StringType()),
        StructField("temp", StringType())
    ]))
])

# Read data from RAW Zone
raw_data = spark.read.schema(raw_schema).parquet(hdfs_raw_path)

# Calculate the previous date
current_date = datetime.now()
current_date_minus_1 = current_date - timedelta(days=1)

# Filter data for the last day
incremental_data = raw_data.filter(unix_timestamp("date") > current_date_minus_1)


processed_data = incremental_data.select(
    col("name"),
    col("date"),
    col("sever-details.server_id").cast("integer").alias("server_id"),
    col("sever-details.location"),
    col("sever-details.temp")
)

# Append processed data in the Processed Zone
processed_data.write.mode("append").partitionBy("date").parquet(hdfs_processed_path)

# Stop the Spark Session
spark.stop()
